### PR TITLE
feat(training): add shared recipe launcher

### PR DESCRIPTION
**Pull request by yaoyu-33**
## Summary

- add a compact `train.sh` + `setup_experiment.py` Slurm entry point with explicit resources, mounts, container, and environment forwarding
- refactor `run_recipe.py` around explicit pretrain/SFT/LoRA/DoRA modes, public dataset names, and a shared training recipe runner
- keep performance launchers and the existing specialized training launchers independent and unchanged
- harden dataset selection, offline text packing, VLM media handling, model auto-registration, and CP finetuning invariants
- keep this change independent of recipe environment-variable configuration; it does not add `ConfigContainer` environment fields or environment-application utilities

## Validation

- `uv run --active --no-sync pre-commit run --all-files`
- 288 focused unit tests in the EOS project container covering the training CLI/setup/runner, dataset helpers, HF adapters, model registration, offline packing, and GPT SFT builder
- final 82-test training CLI/runner replay after the CP-mode guard
- 157 additional focused local tests with one expected skip, including performance entrypoint, packing, CP validation, adapter, and Qwen3-Omni coverage
- all 13 public dataset presets completed a one-iteration EOS smoke run with finite loss
- one-GPU Slurm Qwen3-600M pretrain smoke, one iteration with checkpoint saved
- one-GPU Slurm Qwen3-600M SFT-on-SQuAD smoke, loaded the pretrain checkpoint and saved a checkpoint after one iteration

All Slurm validation jobs completed with exit code 0.

### CODE REVIEW

File: src/megatron/bridge/recipes/utils/dataset_utils.py [hunk 1]
```diff
@@ -124,6 +128,46 @@ def get_blend_fields_from_data_paths(
     "vlm-preloaded",
 ]
 
+PublicDatasetMode = Literal["pretrain", "finetune"]
```
Comment (yaoyu-33) [inline src/megatron/bridge/recipes/utils/dataset_utils.py]: are these meta data only? if so no need to add PublicDatasetMode or PublicDatasetModality. We dont want metadata only info. If it used for data prep, you can keep.

File: src/megatron/bridge/recipes/utils/finetune_utils.py [hunk 2]
```diff
@@ -137,8 +137,10 @@ def default_openmathinstruct2_config(
     pad_seq_to_mult: int = 1,
 ) -> GPTSFTDatasetConfig:
     """Create default OpenMathInstruct-2 dataset configuration for finetuning recipes."""
+    dataset_kwargs = {}
     offline_packing_specs = None
     if packed_sequence:
```
Comment (yaoyu-33) [inline src/megatron/bridge/recipes/utils/finetune_utils.py]: `packed_sequence` means in batch or offline, for all conditions make it very clear.

File: src/megatron/bridge/recipes/utils/finetune_utils.py [hunk 3]
```diff
@@ -137,8 +137,10 @@ def default_openmathinstruct2_config(
...
+        dataset_kwargs["pad_to_max_length"] = True
```
Comment (yaoyu-33) [inline src/megatron/bridge/recipes/utils/finetune_utils.py]: why need to pad to max, I dont see it's necessary.

File: scripts/training/run_recipe.py [hunk 4]
```diff
@@ -12,321 +12,501 @@
...
-TRAIN_FUNCTIONS = {
-    "pretrain": pretrain,
-    "finetune": finetune,
+
+PublicMode = Literal["pretrain", "sft", "lora", "dora"]
+TrainMode = Literal["pretrain", "finetune"]
+
+
+DATASETS = PUBLIC_DATASETS
+LEGACY_DATASET_ALIASES = {
```
Comment (yaoyu-33) [inline scripts/training/run_recipe.py]: no need legacy

File: scripts/training/run_recipe.py [hunk 5]
```diff
@@ -12,321 +12,501 @@
...
+        dest="tokenizer_type",
+        choices=["NullTokenizer", "HuggingFaceTokenizer", "SentencePieceTokenizer"],
     )
-    parser.add_argument(
+    data.add_argument("--tokenizer-model", "--tokenizer_model", dest="tokenizer_model")
+    data.add_argument("--vocab-size", "--vocab_size", type=int, default=32000, dest="vocab_size")
+    data.add_argument(
+        "--offline-packing",
+        "--offline_packing",
+        "--packed-sequence",
```
Comment (yaoyu-33) [inline scripts/training/run_recipe.py]: just offline packing, no packed sequence

File: scripts/training/run_recipe.py [hunk 6]
```diff
@@ -12,321 +12,501 @@
...
+            ("--hf-processor-path", args.hf_processor_path is not None),
+        )
+        if is_selected
+    ]
+    if args.dataset is None:
+        if selected_options:
+            raise ValueError(f"{', '.join(selected_options)} require --dataset.")
+        return
+
+    if args.dataset in LEGACY_DATASETS:
```
Comment (yaoyu-33) [inline scripts/training/run_recipe.py]: no need LEGACY_DATASETS

File: scripts/training/run_recipe.py [hunk 7]
```diff
@@ -12,321 +12,501 @@
...
+        "--test-data-path",
+        "--test_data_path",
+        dest="test_data_path",
+        help="Local VLM test JSON or JSONL path.",
+    )
+    data.add_argument(
+        "--media-root",
+        "--media_root",
+        dest="media_root",
+        help="Root for relative local media paths or LLaVA-Video assets.",
```
Comment (yaoyu-33) [inline scripts/training/run_recipe.py]: not good if only specific to one of the dataset.

File: scripts/performance/setup_experiment.py [hunk 8]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env python3
```
Comment (yaoyu-33) [inline scripts/performance/setup_experiment.py:1]: revert this file; no need.

File: docs/training/entry-points.md [hunk 9]
```diff
@@ -8,7 +8,14 @@ Use `pretrain()` for language-model pretraining jobs that use `GPTDatasetConfig`
 
 Use `finetune()` for full SFT and PEFT. The function validates that either `checkpoint.pretrained_checkpoint` or `checkpoint.load` is set, then calls the same underlying training loop used by `pretrain()`. PEFT does not use a separate entry point: set `cfg.peft` to a LoRA or DoRA config, use `GPTSFTDatasetConfig`, `DirectHFSFTDatasetConfig`, or a specialized dataset provider, and launch through `finetune()`.
 
-The generic recipe launcher, `scripts/training/run_recipe.py`, follows the same split. Dataset types beginning with `llm-pretrain` run `pretrain()`. SFT, PEFT, VLM, and diffusion fine-tuning dataset types run `finetune()`.
+The public Slurm launcher is `scripts/training/train.sh`, backed by `setup_experiment.py`, which executes
+`run_recipe.py` inside the training container. Select exactly one model stem with `--model` or one complete library
+recipe function with `--recipe`, and always pass `--mode pretrain`, `sft`, `lora`, or `dora`. Public dataset names
+such as `megatron-indexed`, `openmathinstruct2`, `local-jsonl`, `local-vlm`, `medpix`, and `mock` resolve to the
```
Comment (cuichenx) [inline docs/training/entry-points.md]: megatron-indexed, local-jsonl and local-vlm are formats while the other ones are dataset names, they should be talked about separately?

File: scripts/performance/run_script.py [hunk 10]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
+# Copyright (c) 2026, NVIDIA CORPORATION.  All rights reserved.
```
Comment (cuichenx) [inline scripts/performance/run_script.py]: should not change

File: scripts/training/README.md [hunk 11]
```diff
@@ -1,275 +1,202 @@
...
-
-# Gemma
-uv run python -m torch.distributed.run --nproc_per_node=8 run_recipe.py --recipe gemma3_1b_pretrain_1gpu_h100_bf16_config
-
-# Qwen
-uv run python -m torch.distributed.run --nproc_per_node=8 run_recipe.py --recipe qwen3_8b_pretrain_4gpu_h100_bf16_config
-
-# GPT
-uv run python -m torch.distributed.run --nproc_per_node=8 run_recipe.py --recipe vanilla_gpt_pretrain_1gpu_h100_bf16_config
+./scripts/training/train.sh \
```
Comment (cuichenx) [inline scripts/training/README.md:27]: all model examples should adopt this new entry point?

File: scripts/training/recipe_runner.py
Comment (cuichenx) [inline scripts/training/recipe_runner.py:1]: these helper functions should be placed in src/ not scripts

File: scripts/training/run_recipe.py [hunk 12]
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env python3
-# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
+# Copyright (c) 2026, NVIDIA CORPORATION.  All rights reserved.
```
Comment (cuichenx) [inline scripts/training/run_recipe.py]: should not change

File: scripts/training/run_recipe.py [hunk 13]
```diff
@@ -12,321 +12,501 @@
...
-        help="Recipe function name (e.g., llama32_1b_pretrain_config, gemma3_1b_sft_config, gemma3_1b_peft_config)",
+
+    recipe_args = parser.add_argument_group("Recipe construction")
+    recipe_args.add_argument("--seq-length", "--seq_length", type=int, dest="seq_length")
+    recipe_args.add_argument("--hf-path", "--hf_path", dest="hf_path")
+
+    training = parser.add_argument_group("Training")
+    training.add_argument("--max-steps", "--max_steps", type=int, dest="max_steps")
+    training.add_argument("--global-batch-size", "--global_batch_size", type=int, dest="global_batch_size")
+    training.add_argument("--micro-batch-size", "--micro_batch_size", type=int, dest="micro_batch_size")
```
Comment (cuichenx) [inline scripts/training/run_recipe.py]: a lot of things like micro batch size can be changed either through hydra config override, so there's no need to create argparse arguments for them?

File: src/megatron/bridge/recipes/utils/dataset_utils.py (hunk 1 above)
Comment (yaoyu-33) [reply src/megatron/bridge/recipes/utils/dataset_utils.py]: Removed the standalone `PublicDatasetMode` and `PublicDatasetModality` aliases. The operational mode and modality fields remain inline on `PublicDatasetSpec` because launcher validation and dataset application use them.

File: scripts/training/run_recipe.py (hunk 7 above)
Comment (yaoyu-33) [reply scripts/training/run_recipe.py]: Generalized `--media-root` to resolve relative image and video paths. It is now used by both `local-vlm` and `llava-video-178k`, with coverage for both.

File: src/megatron/bridge/recipes/utils/finetune_utils.py (hunk 2 above)
Comment (yaoyu-33) [reply src/megatron/bridge/recipes/utils/finetune_utils.py]: Clarified every `packed_sequence` parameter docstring: it enables offline text SFT packing and does not enable VLM in-batch packing.

File: scripts/training/run_recipe.py (hunk 6 above)
Comment (yaoyu-33) [reply scripts/training/run_recipe.py]: Removed `LEGACY_DATASETS`, its validation/application branches, and legacy imports. Tests now verify all former names are rejected.

File: scripts/performance/setup_experiment.py (hunk 8 above)
Comment (yaoyu-33) [reply scripts/performance/setup_experiment.py:1]: Reverted this performance integration. `setup_experiment.py`, `run_script.py`, and the related performance entrypoint test now match the pre-PR base byte-for-byte.

File: scripts/training/run_recipe.py (hunk 5 above)
Comment (yaoyu-33) [reply scripts/training/run_recipe.py]: Removed the `--packed-sequence` aliases. Only `--offline-packing` and `--offline_packing` remain.

File: src/megatron/bridge/recipes/utils/finetune_utils.py (hunk 3 above)
Comment (yaoyu-33) [reply src/megatron/bridge/recipes/utils/finetune_utils.py]: Removed the newly added forced `pad_to_max_length` setting from OpenMathInstruct-2, GSM8K, thinking, and local JSONL offline-packing paths, with regression assertions.

File: scripts/training/run_recipe.py (hunk 4 above)
Comment (yaoyu-33) [reply scripts/training/run_recipe.py]: Removed `LEGACY_DATASET_ALIASES` and the compatibility-resolution path. The launcher now accepts only the public dataset names.

File: scripts/training/setup_experiment.py
Comment (cuichenx) [inline scripts/training/setup_experiment.py:1]: do we need to keep both setup_experiment.py scripts?

File: src/megatron/bridge/recipes/utils/dataset_utils.py [hunk 14]
```diff
@@ -124,6 +128,46 @@ def get_blend_fields_from_data_paths(
...
+        train_mode="finetune",
+        modality="vlm",
+        hf_dataset_name="llava_video_178k",
+    ),
+    "medpix": PublicDatasetSpec(train_mode="finetune", modality="vlm", hf_dataset_name="medpix"),
+    "raven": PublicDatasetSpec(train_mode="finetune", modality="vlm", hf_dataset_name="raven"),
+    "rdr": PublicDatasetSpec(train_mode="finetune", modality="vlm", hf_dataset_name="rdr"),
+}
+
+PUBLIC_HF_VLM_DATASETS = {
```
Comment (cuichenx) [inline src/megatron/bridge/recipes/utils/dataset_utils.py]: is this necessary?

**Review by cuichenx — COMMENTED** [review]
need to update the recipe recommender skill also

File: scripts/training/setup_experiment.py
Comment (yaoyu-33) [reply scripts/training/setup_experiment.py:1]: I cannot modify them at the same time,  can only do training folder first.

File: scripts/training/run_recipe.py (hunk 13 above)
Comment (yaoyu-33) [reply scripts/training/run_recipe.py]: agree

File: scripts/training/README.md (hunk 11 above)
Comment (yaoyu-33) [reply scripts/training/README.md:27]: yes

File: scripts/training/recipe_runner.py
Comment (yaoyu-33) [reply scripts/training/recipe_runner.py:1]: right now it's only used for run recipe, might be okay? should not be used in other cases.

**Comment by yaoyu-33** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/4856/commits/c9da8261ea0d2320e15e0a7aeb5c8cde0325e9af
